File: src/main/java/com/matamercer/microblog/services/PostTagService.java
package com.matamercer.microblog.services;

import com.matamercer.microblog.models.entities.Blog;
import com.matamercer.microblog.models.entities.PostTag;
import com.matamercer.microblog.models.repositories.PostTagRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class PostTagService {

    @Autowired
    private PostTagRepository postTagRepository;

    public PostTag findOrCreateByName(String name) {
        PostTag postTag = postTagRepository.findByName(name);
        if (postTag == null) {
            postTag = postTagRepository.save(new PostTag(name));
        }
        return postTag;
    }

    public PostTag getTag(String name) {
        return postTagRepository.findByName(name);
    }

    public Set<PostTag> getTags(List<String> postTagNames){
        return postTagNames.stream().map(this::getTag).collect(Collectors.toSet());
    }

    public Map<PostTag, Integer> getTopTagsByPosts(Blog blog, int page, int pageSize){
        return postTagRepository.findByBlogSortedByMostUsedMap(blog, PageRequest.of(page, pageSize, Sort.Direction.ASC, "postTagCount"));
    }

    public void deleteTag(PostTag tag) {
        postTagRepository.delete(tag);
    }
}

File: src/main/java/com/matamercer/microblog/services/PostTagService.kt
package com.matamercer.microblog.services

import com.matamercer.microblog.Exceptions.NotFoundException
import com.matamercer.microblog.models.entities.Blog
import com.matamercer.microblog.models.entities.PostTag
import com.matamercer.microblog.models.repositories.PostTagRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import java.util.stream.Collectors

@Service
class PostTagService @Autowired constructor(private val postTagRepository: PostTagRepository) {
    fun findOrCreateByName(name: String): PostTag {
        return postTagRepository.findByName(name) ?: postTagRepository.save(PostTag(name))
    }

    fun getTag(name: String): PostTag {
        return postTagRepository.findByName(name) ?: throw NotFoundException("Post tag not found")
    }

    fun getTags(postTagNames: List<String>): Set<PostTag> {
        return postTagNames.stream().map { name: String -> getTag(name) }.collect(Collectors.toSet())
    }

    fun getTopTagsByPosts(blog: Blog, page: Int, pageSize: Int): Map<PostTag, Int>? {
        return findByBlogSortedByMostUsedMap(
            blog,
            PageRequest.of(page, pageSize, Sort.Direction.ASC, "postTagCount")
        )
    }

    fun findByBlogSortedByMostUsedMap(blog: Blog, pageRequest: Pageable?): Map<PostTag, Int>? {
        return postTagRepository.findByBlogSortedByMostUsed(blog, pageRequest).toList().associate { it.postTag to it.postTagCount }
            .toList().sortedBy { it.second }.toMap()
    }

    fun deleteTag(tag: PostTag) {
        postTagRepository.delete(tag)
    }
}